src/mbgl/sprite/sprite_atlas.cpp

```diff
@@ -115,180 +132,161 @@ void SpriteAtlas::setSprite(const std::string& name, std::shared_ptr<const Sprit
-    Rect<uint16_t> rect = bin.allocate(pack_width, pack_height);
-    if (rect.w == 0) {
-        return rect;
+    auto it = entries.find(name);
+    if (it == entries.end()) {
+        if (!entries.empty()) {
+            Log::Info(Event::Sprite, "Can't find sprite named '%s'", name.c_str());
+        }
+        return {};
     }
 
-    return rect;
-}
-
-optional<SpriteAtlasElement> SpriteAtlas::getImage(const std::string& name,
-                                                   const SpritePatternMode mode) {
-    std::lock_guard<std::recursive_mutex> lock(mtx);
+    Entry& entry = it->second;
 
-    auto rect_it = images.find({ name, mode });
-    if (rect_it != images.end()) {
-        return SpriteAtlasElement { rect_it->second.pos, rect_it->second.spriteImage, rect_it->second.spriteImage->pixelRatio / pixelRatio };
+    if (mode == SpritePatternMode::Single && entry.iconRect) {
+        return SpriteAtlasElement {
+            *entry.iconRect,
+            entry.spriteImage,
+            size,
+            pixelRatio
+        };
     }
 
-    auto sprite = getSprite(name);
-    if (!sprite) {
-        return {};
+    if (mode == SpritePatternMode::Repeating && entry.patternRect) {
+        return SpriteAtlasElement {
+            *entry.patternRect,
+            entry.spriteImage,
+            size,
+            pixelRatio
+        };
     }
 
-    Rect<uint16_t> rect = allocateImage(*sprite);
+    const uint16_t pixelWidth = std::ceil(entry.spriteImage->image.size.width / pixelRatio);
+    const uint16_t pixelHeight = std::ceil(entry.spriteImage->image.size.height / pixelRatio);
+
+    // Increase to next number divisible by 4, but at least 1.
+    // This is so we can scale down the texture coordinates and pack them
+    // into 2 bytes rather than 4 bytes.
+    const uint16_t packWidth = (pixelWidth + 1) + (4 - (pixelWidth + 1) % 4);
+    const uint16_t packHeight = (pixelHeight + 1) + (4 - (pixelHeight + 1) % 4);
+
+    // We have to allocate a new area in the bin, and store an empty image in it.
+    Rect<uint16_t> rect = bin.allocate(packWidth, packHeight);
     if (rect.w == 0) {
         if (debug::spriteWarnings) {
             Log::Warning(Event::Sprite, "sprite atlas bitmap overflow");
         }
         return {};
     }
 
-    const Holder& holder = images.emplace(Key{ name, mode }, Holder{ sprite, rect }).first->second;
-    copy(holder, mode);
-
-    return SpriteAtlasElement { rect, sprite, sprite->pixelRatio / pixelRatio };
-}
-
-optional<SpriteAtlasPosition> SpriteAtlas::getPosition(const std::string& name,
-                                                       const SpritePatternMode mode) {
-    std::lock_guard<std::recursive_mutex> lock(mtx);
+    copy(entry.spriteImage->image, rect, mode);
 
-    auto img = getImage(name, mode);
-    if (!img) {
-        return {};
+    if (mode == SpritePatternMode::Single) {
+        entry.iconRect = rect;
+    } else {
+        entry.patternRect = rect;
```

Comment:
It's possible both rects are needed simultaneously -- I've clarified this in the comment.